### ISSUE TITLE
NATS/Jetstream server does not expire messges fast enough from the filestore when stream size is very large. 

### ISSUE THREAD

**Issue by jghelto** [nats-io/nats-server#4293]
## Defect

NATS/Jetstream server does not expire messages fast enough from the filestore when stream size is very large. Stream size continues to grow and old messages continue to get further and further in the past.  

Make sure that these boxes are checked before submitting your issue -- thank you!

 - [X ] Included `nats-server -DV` output
 - [X] Included a [Minimal, Complete, and Verifiable example] (https://stackoverflow.com/help/mcve)

#### Versions of `nats-server` and affected client libraries used:

nats-server-2.10.0-beta.42
nats-cli version 0.0.35

#### OS/Container environment:

synadia/nats-server:nightly, running in a docker container (not clustered) on an HP SuperMicro SuperStorage server with 256GB of RAM and 88 CPU cores.  The container is using File storage for persistence.  The file storage is a RAID 0 (6 HDD @ 7200RPM) total of 76 TB. 

#### Steps or code to reproduce the issue:

1. Start up Docker container using this command. docker run -d --name test -p 4222:4222 -p 6222:6222 -p 8222:8222 -v /data3:/data3 synadia/nats-server:nightly -js -sd=/data3
2. Create a stream with a single subject.  Set the limits for a max-age of 3 days (I suspect 2 days would work as well).    nats stream add --config=./config.json  (config.json) attached. In my case, I'm creating 8 identical streams each with a different name.  Each stream has a single subject of the same name as the stream. Note, I  have acks turned off. My publishers are adhering to the NATS core API and publishing to a subject only.  My server uses Jetstream to persist the messages and allow me to replay them on demand. 
3. Start pushing data to the stream.  In my case, I'm pushing ~1.45 GiB/minute to the stream in 400KB messages. This rounds out to 81 GiB/hour and ~2.1 GiB/day.  I need to be able to keep 3 days worth of messages. 

#### Expected result:

At the 3d0H0M0S mark the old messages should be purged where so the first message is always 3 days older than the most recent message.

#### Actual result:

At the 3 day mark, old messages are purged, but they are purged very slowly.  At 3 days and 4 hours, the oldest message was 3 days and 1 hour old. After 5 days, my oldest message is now 4 days old. Stream info attached. 

NOTE: I ran this same test with a 1 day retention policy and the data was purged on time as expected. This seems to be an issue with the message expiration keeping up on very large, file-based streams. 

[nats_dv.txt](
[stream_info.txt](https://github.com/nats-io/nats-server/files/11996479/stream_info.txt)
https://github.com/nats-io/nats-server/files/11996451/nats_dv.txt)
[config_json.txt](https://github.com/nats-io/nats-server/files/11996470/config_json.txt)
[stream_info.txt](https://github.com/nats-io/nats-server/files/11996483/stream_info.txt)

**Comment by derekcollison** [comment]
Do you see warnings in the server logs around `lag` getting too far behind? In general we do not recommend using high performing NATS core into JetStream streams and in future versions of the server we may discard those messages from the JetStream layer.

Also could you clarify numbers for this one?

```
3. Start pushing data to the stream. In my case, I'm pushing ~1.45 GiB/minute to the stream in 400KB messages. This rounds out to 81 GiB/hour and ~2.1 GiB/day. I need to be able to keep 3 days worth of messages.
```

**Comment by jghelto** [comment]
The logs don't show anything until I tried to do a purge from the CLI about 3 hours after the 3 day mark.  You can see there the internal subscriptions took 2 seconds to get stream info, then took an exceptionally long time to purge (again from the command line). The log is attached. 

On the data numbers, that should have been 1.45GiB/min, 81 GiB/hour and 2.1TiB/day (not 2.1 GiB/day). Sorry for the confusion.  

I'm actually pushing to the stream AND subject, and not just the subject as I described earlier.  I have just turned acks off on the stream because it was causing me to drop data.  I am using a C++ client and was using async acks but was still dropping data.  In my case, I'm receiving messages over multicast and publishing them to Jetstream for near real-time consumers to have access to them without having to access the host network and subscribe to multicast. Jetstream also provides the added benefit of persistence and message replay. I was dropping multicast packets when I had async acks on.  That's potentially something I can fix later.  Right now I just need to see if Jetstream can solve my initial problem, which is to provide a translation layer from multicast to container networking, and to provide a persistent message store and replay system. It's so close! 

[nats_log.txt](https://github.com/nats-io/nats-server/files/12008927/nats_log.txt)

**Comment by derekcollison** [comment]
What does `nats stream info TEST` show?

**Comment by jghelto** [comment]
nats stream info TEST
Information for Stream TEST created 2023-07-04 15:28:22

             Subjects: TEST
             Replicas: 1
              Storage: File

Options:

            Retention: Limits
     Acknowledgements: false
       Discard Policy: Old
     Duplicate Window: 30s
    Allows Msg Delete: true
         Allows Purge: true
       Allows Rollups: false

Limits:

     Maximum Messages: unlimited
  Maximum Per Subject: unlimited
        Maximum Bytes: unlimited
          Maximum Age: 3d0h0m0s
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited


State:

             Messages: 19,252,792
                Bytes: 7.2 TiB
             FirstSeq: 8,714,403 @ 2023-07-06T12:51:09 UTC
              LastSeq: 27,967,194 @ 2023-07-10T23:57:49 UTC
     Active Consumers: 0
   Number of Subjects: 1

**Comment by derekcollison** [comment]
And you are running a single nats-server ?

**Comment by jghelto** [comment]
Yes.  8 streams on the same server. This does work when retention policy is 1 day.

**Comment by derekcollison** [comment]
Could you share a `tree -h` output for the storage directory?

**Comment by jghelto** [comment]
Unfortunately I had to stop it and clean up the disk as I was nearly out of space.  I'm about to start up a new instance so I can grab a tree of the storage directory at some point after that.  I assume you want it when I reach the max-age?

**Comment by jghelto** [comment]
I ran a slightly different test where I had multiple Jetstream instances, each with only 2-3 streams (same size as before) on them.  Some of the streams are keeping up with expiring messages, while some are not.  Generally, all streams on a single broker are either keeping up or not. I am working on getting you a tree for a broker that is working versus one that is not. Obviously the tree command is taking some time to run.  As soon as I get the tree, I will upload the two files. Each Jetstream instance is on its own RAID 0.  I'll provide two examples from different HP DL360 servers.

**Comment by derekcollison** [comment]
Thanks that would be helpful.

**Comment by jghelto** [comment]
[server_bad_part_1.txt.gz](https://github.com/nats-io/nats-server/files/12075074/server_bad_part_1.txt.gz)
[server_bad_part_2.txt.gz](https://github.com/nats-io/nats-server/files/12075078/server_bad_part_2.txt.gz)
[server_good.txt.gz](https://github.com/nats-io/nats-server/files/12075079/server_good.txt.gz)

I've attached three files.  They are server_bad_part_1.txt.gz, server_bad_part2.txt.gz and server_good.txt.gz.  The *bad* files were split using the Linux split command because even after gzipping the original file it was too large to upload.  These files all contain the output of the tree command. The server_bad ones from the server not keeping up with expiring messages while server_good is currently keeping up. In each case there is a single NATS/Jetstream instance running on as a Docker container on a separate DL360 server each and each with it's own 6 HDD (7200 RPM) RAID 0 disk. Each instance is serving 3 message producers at the rate I posted earlier. Basically I split the message producers into multiple Jetstream instances to see if that made a difference. Note that there are longhorn.replica volumes mounted on both disks. I previously ran Longhorn on these servers and Longhorn had been scheduling PVCs on these nodes. It is not now, but the longhorn.replica volumes remain.

**Comment by derekcollison** [comment]
Possible to run without docker and longhorn but on same HW?

**Comment by jghelto** [comment]
I should be able to do that. I’ll let you know how that goes

**Comment by derekcollison** [comment]
I think the 2.10 server should handle now well as long as you were not simply over publishing and causing huge lags in the server between the app publish and when the server can actually store and process the messages.

**Comment by derekcollison** [comment]
We will complain when running in clustered mode if sending to a R>1 stream. On single server we may not complain, but you can tell by watching the last msg you published (if under test you start at sequence 1 with new stream) and what stream info reports.

So if your application says I just sent msg 1,000,000 and stream info says the stream has 400k msgs in it the server has staged but not processed ~600k msgs.
